components/ai_chat/core/browser/ai_chat_service.cc

```diff
@@ -0,0 +1,373 @@
+
+  return conversation;
+}
+
+void AIChatService::MarkAgreementAccepted() {
+  SetUserOptedIn(profile_prefs_, true);
+}
+
+void AIChatService::GetActionMenuList(GetActionMenuListCallback callback) {
+  std::move(callback).Run(ai_chat::GetActionMenuList());
+}
+
+void AIChatService::GetPremiumStatus(GetPremiumStatusCallback callback) {
+  credential_manager_->GetPremiumStatus(
+      base::BindOnce(&AIChatService::OnPremiumStatusReceived,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+}
+
+void AIChatService::GetCanShowPremiumPrompt(
+    GetCanShowPremiumPromptCallback callback) {
+  bool has_user_dismissed_prompt =
+      profile_prefs_->GetBoolean(prefs::kUserDismissedPremiumPrompt);
+
+  if (has_user_dismissed_prompt) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  base::Time last_accepted_disclaimer =
+      profile_prefs_->GetTime(prefs::kLastAcceptedDisclaimer);
+
+  // Can't show if we haven't accepted disclaimer yet
+  if (last_accepted_disclaimer.is_null()) {
+    std::move(callback).Run(false);
+    return;
+  }
+
+  base::Time time_1_day_ago = base::Time::Now() - base::Days(1);
+  bool is_more_than_24h_since_last_seen =
+      last_accepted_disclaimer < time_1_day_ago;
+
+  if (is_more_than_24h_since_last_seen) {
+    std::move(callback).Run(true);
+    return;
+  }
+
+  std::move(callback).Run(false);
+}
+
+void AIChatService::DismissPremiumPrompt() {
+  profile_prefs_->SetBoolean(prefs::kUserDismissedPremiumPrompt, true);
+}
+
+void AIChatService::OnPremiumStatusReceived(GetPremiumStatusCallback callback,
+                                            mojom::PremiumStatus status,
+                                            mojom::PremiumInfoPtr info) {
+#if BUILDFLAG(IS_ANDROID)
+  // There is no UI for android to "refresh" with an iAP - we are likely still
+  // authenticating after first iAP, so we should show as active.
+  if (status == mojom::PremiumStatus::ActiveDisconnected &&
+      profile_prefs_->GetBoolean(prefs::kBraveChatSubscriptionActiveAndroid)) {
+    status = mojom::PremiumStatus::Active;
+  }
+#endif
+
+  last_premium_status_ = status;
+  if (ai_chat::HasUserOptedIn(profile_prefs_) && ai_chat_metrics_ != nullptr) {
+    ai_chat_metrics_->OnPremiumStatusUpdated(false, status, std::move(info));
+  }
+  model_service_->OnPremiumStatus(status);
+  std::move(callback).Run(status, std::move(info));
+}
+
+void AIChatService::MaybeEraseConversation(
+    ConversationHandler* conversation_handler) {
+  if (!conversation_handler->IsAnyClientConnected() &&
+      (!features::IsAIChatHistoryEnabled() ||
+       !conversation_handler->HasAnyHistory())) {
+    // Can erase because no active UI and no history, so it's
+    // not a real / persistable conversation
+    auto uuid = conversation_handler->get_conversation_uuid();
+    conversation_observations_.RemoveObservation(conversation_handler);
+    conversation_handlers_.erase(uuid);
```

Comment:
Not sure if we should also update `content_conversations_`, it can be replaced when a failed lookup happend in `GetOrCreateConversationHandlerForContent` but if it doesn't get accessed for a long time or even content no longer exists, the entry would stay in memory forever until shutdown.

Reply:
Yes we can erase any entries in the NavigationID to ConversationID map here. But in general there is a problem with that map (behind the AIChatConversationHistory feature flag) in that when the content for a conversation is navigated away we don't know how long to keep the entry in that map for in order to support matching the conversation on back-navigation. I think it's probably ok to keep it until the Tab closes, whilst the navigation ID is alive somewhere in the original WebContents navigation entry history because I think when a Tab is re-opened, its history items get different IDs than the original WebContents. And it's not much data.

Reply:
For now, I've added erasing `content_conversations_` based on values when a conversation is erased. We'll have to handle when content is erased in https://github.com/brave/brave-core/pull/25339